in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+      .setArgsOfCalls([
+        ...appPayload.packedArguments,
+        ...feePayload.packedArguments,
+         entrypointPackedArgs
+      ])
+      .addAuthWitness(
+        await this.auth.createAuthWit(appPayload.hash())
+      )
+      .addAuthWitness(
+        await this.auth.createAuthWit(setupPayload.hash())
+      )
+  }
+
+  private getEntrypointAbi() {
+    return {
+      name: 'entrypoint',
+      isInitializer: false,
+      // ... same as before
+    }
+  }
+}
+
+```
+
+#### Fill in the `TxExecutionRequest`
+
+The abstract `BaseWallet` can implement:
+
+```ts
+async getTxExecutionRequest(userRequest: UserRequest): Promise<TxExecutionRequest> {
+  if (!userRequest.gasSettings) {
+    throw new Error('Gas settings must be provided');
+  }
+  if (!userRequest.paymentMethod) {
+    throw new Error('Payment method must be provided');
+  }
+
+  const builder = new TxExecutionRequestBuilder();
+
+  for (const call of request.calls) {
+    addMainFunctionCall(builder, call);
+    if (call.deploymentOptions?.registerClass) {
+      addContractClassRegistration(builder, call);
+    }
+    if (call.deploymentOptions?.publicDeploy) {
+      addPublicContractDeployment(builder, call);
+    }
+    // if the user is giving us an artifact,
+    // allow the PXE to access it
+    if (call.contractArtifact) {
+      builder.addTransientContract({
+        artifact: call.contractArtifact,
+        instance: call.contractInstance,
+      });
+    }
+  }
+
+  // Add stuff needed for setup, e.g. function calls, auth witnesses, etc.
+  await userRequest.paymentMethod.adaptTxExecutionRequest(builder, userRequest);
+
+  // Adapt the request to the entrypoint in use.
+  // Since BaseWallet is abstract, this will be implemented by the concrete class.
+  this.adaptTxExecutionRequest(builder, userRequest);
+
+  return builder.build();
+
+}
+```
+
+### BaseWallet.#simulateInner
+
+```ts
+// Used by simulate and read
+async #simulateInner(userRequest: UserRequest): ReturnType<BaseWallet['simulate']> {
+  const txExecutionRequest = await this.getTxExecutionRequest(userRequest);
+  const simulatedTx = await this.simulateTx(txExecutionRequest, builder.simulatePublicFunctions, builder.from); 
+  const decodedReturn = decodeSimulatedTx(simulatedTx, builder.functionAbi);
+  return {
+    tx: simulatedTx.tx,
+    publicOutput: simulatedTx.publicOutput,
+    privateOutput: simulatedTx.privateReturnValues,
+    executionResult: simulatedTx.executionResult,
+    result: decodedReturn,
+    request: userRequest,
+  };
+}
+```
+
+### BaseWallet.simulate
+
+```ts
+
+async simulate(userRequest: UserRequest): {
+    tx: simulatedTx.tx,
+    publicOutput: simulatedTx.publicOutput,
+    privateOutput: simulatedTx.privateReturnValues,
+    executionResult: simulatedTx.executionResult,
+    result: decodedReturn,
+    request: initRequest,
+} {
+  // If we're simulating, we need to have the payment method set.
+  // Users should use `read` if they just want to see the result.
```

Comment:
It is not clear to me that I would be using a "read" if I just want the results when it is not a `view` function. For `viem` I think the `read` is only used for `view` and `pure` functions? It is not uncommon to have something like a simulate a real call that is not static but still to get some return values from it, so seems slightly annoying we we have to deal with a payment method for something like that?

Reply:
You're right that viem's read is only view/pure functions, and I agree there there is some clunkiness here to iron out.

Do you think users should be able to simulate if they have no means of actually paying for what they want to simulate?

Reply:
I have forgotten, with changes on fees and such, at what point do we really require the fees to be paid, if it is only the base rollup stage, then you should be able to simulate it entirely without dealing with the fees?